Begin refactoring standard libraries

cgranade opened the pull request:
This PR begins to separate out the different namespaces in the standard libraries, and to refactor them to be more consistent..

### TODO ###

- [ ] Finish splitting out namespaces
  - [x] Split out Qecc
  - [x] Split out amplitude amplificaiton
  - [x] Split out characterization
- [ ] Remove BE from arithmetic
- [ ] Apply consistent naming:
    - [x] type conversion functions.
    - [ ] oracle representation conversions
- [x] Update deprecated syntax (mostly logical operators)
- [x] Update tests to use new names
- [ ] Update relevant README.md files
- [ ] Apply style guide within new namespaces
    - [ ] amplitude amplification (attn @guanghaolow)

martinquantum commented:
This for spearheading the refactoring and starting this branch, @cgranade! I think this looks great. Two items for feedback about the Standard package: 1. I like the way Arithmetic, Preparation, and Simulation are broken out into subfolders (and how it is reflected in subnamespaces, e.g., Microsoft.Quantum.Preparation). My question is whether we should also break out PhaseEstimation, QCVV, AmplitudeAmplification, and QECC in the same way? (and leave Arrays, Asserts, Combinators, DataStructures, Enumeration, Math, and Utils in the Canon). 2. I think it'd be best to have a global default of little endian for the arithmetic functions and operations. So it would be good to indicate explicitly (as a suffix) wherever there is another convention (e.g. BE), but otherwise to simplify the function/operations names by dropping the trailing LE.

cgranade commented:
Thanks for the feedback, @martinquantum! I agree that breaking things down further would make a lot of sense; my intent with making this a draft PR was to prototype that work by using arithmetic as a test case, but I agree, moving more of the specialized functionality into their own namespaces can make it easier to find relevant functionality. The counter to that is that we don't want users to necessarily have to have lots of additional `open` statements, such that identifying high-level functional areas makes a lot of sense.

At the moment, the subfolders within the existing `Canon/` folder help identify that as per your comment, such that we will likely have something a bit like the following:

- Microsoft.Quantum.Standard.nupkg
  - namespace Microsoft.Quantum.Canon
  - namespace Microsoft.Quantum.Arithmetic
  - namespace Microsoft.Quantum.AmplitudeAmplification
  - namespace Microsoft.Quantum.Preparation
  - namespace Microsoft.Quantum.Simulation
  - namespace Microsoft.Quantum.ErrorCorrection (was Qecc)

The ones I'm a bit less sure of the best way to handle are PhaseEstimation and Qcvv. I could see an argument for including both into a new Microsoft.Quantum.Characterization namespace, reflecting the commonality there.

With respect to ②, I'll add that to the TODO list for this PR  (editing the root comment) so as to not lose track of that I still need to make that change.

cgranade commented:
There is definitely still work to be done (most notably, updating README and `///` comments, updating implementations to use newly refactored namespaces, and updating the amplitude amplification feature), but I think this PR is now in a state where it gives us something we can start to play with and work on while I finish refactoring and addressing those TODOs.

cgranade commented:
I have now rebased this PR onto the most recent master, so it should be ready for review. Thanks!

cgranade commented:
> Thanks a lot for working on this @cgranade! I think it looks great.
> As an additional comment, I'm not so sure about replacing `Assert` with `Claim` because the latter (to me) does not sounds like the claim will be checked for correctness; and since we are using it in all the tests, this is a bit unfortunate.

Thanks for the review, @thomashaener! I'll work on addressing your feedback and should have the PR updated shortly. As for "Assert" versus "Claim," my intent there was to move away from using verbs to describe functions; taking that view, a claim is either a value of type Unit (that is, is a valid claim), or is a value of type ⊥. Since no value of type ⊥ exists or can exist, the program halts if the claim is not valid. 

That said, I'm happy to revisit the terminology there and to apply updates to the new style guide accordingly. One thought might be to refer to an `EqualityAssertion` as a noun, rather than to `AssertEquality` as a verb. That would clearly distinguish them from diagnostic operations like `AssertQubitState` that remain named as verbs.

cgranade commented on Standard/src/Measurement/Integer.qs:
```diff
@@ -0,0 +1,60 @@
+    ///
+    /// # See Also
+    /// - Microsoft.Quantum.Canon.MeasureIntegerBE
+    operation MeasureIntegerLE(target : LittleEndian) : Int {
```
I went back and forth on that a lot, actually. I think it comes down to how much we think of arithmetic as a basic part of the standard libraries, and how many other UDTs we expect to have specialized measurement operations down the road.

cgranade replied:
Resolved by moving back to Microsoft.Quantum.Arithmetic.

cgranade commented on Standard/src/Characterization/ProcessTomography.qs:
```diff
@@ -0,0 +1,161 @@
+
+    /// # Summary
+	/// Jointly measures a register of qubits in the Pauli Z basis.
+	///
```
Thanks for the catch! There's a lot of that still in the libraries from the various reformattings that we've done along the way; I've tried to find them as I went along, but I definitely didn't catch everything, so I appreciate the help.

cgranade commented on Standard/src/Arithmetic/Shorthand.qs:
```diff
@@ -0,0 +1,32 @@
+    open Microsoft.Quantum.Canon;
+
+    /// # Summary
+    /// This computes applies the in-place majority operation on 3 qubits.
```
Thanks! I've not fully fixed all docs yet, as that will take another pass, but I appreciate catching out the especially egregious examples like this one.

cgranade commented on Standard/src/Arithmetic/Data.qs:
```diff
@@ -0,0 +1,38 @@
+// Licensed under the MIT License.
+
+namespace Microsoft.Quantum.Arithmetic {
+
```
Done, thanks!

cgranade commented:
> > Thanks a lot for working on this @cgranade! I think it looks great.
> > As an additional comment, I'm not so sure about replacing `Assert` with `Claim` because the latter (to me) does not sounds like the claim will be checked for correctness; and since we are using it in all the tests, this is a bit unfortunate.
> 
> Thanks for the review, @thomashaener! I'll work on addressing your feedback and should have the PR updated shortly. As for "Assert" versus "Claim," my intent there was to move away from using verbs to describe functions; taking that view, a claim is either a value of type Unit (that is, is a valid claim), or is a value of type ⊥. Since no value of type ⊥ exists or can exist, the program halts if the claim is not valid.
> 
> That said, I'm happy to revisit the terminology there and to apply updates to the new style guide accordingly. One thought might be to refer to an `EqualityAssertion` as a noun, rather than to `AssertEquality` as a verb. That would clearly distinguish them from diagnostic operations like `AssertQubitState` that remain named as verbs.

I've now modified "Claim" to be "Fact", in line with Xunit and similar packages. This has the advantage of being even more strongly identifiable as a noun than "Claim".
